Increase usable scrolling range in wxMSW by a factor of 10,000

vadz opened the pull request:
This changes setting the DC origin, and hence scrolling, which is based on it, to be done entirely inside wx itself instead of using Win32 GDI functions for doing it. The reason is that GDI is limited to 2^27 device space and this is not enough for more than ~5,000,000 items which is perfectly achievable in controls such as `wxDataViewCtrl` (see #17550) or `wxGrid` (can be tested after applying the first commit here and entering 10,000,000 for the virtual grid size in the dialog shown by the "File|Virtual table test" menu command).

The changes themselves are reasonably simple and seem to work for me, but they're, of course, quite dangerous as there may be some hidden assumptions relying on the DC origin being actually shifted somewhere that I'm missing. I'm _not_ sure if applying this so close to 3.2.0 is a good idea, but OTOH #17550 is a real problem (and there were other reports of similar problems with `wxGrid` too in the past) and I don't want to apply the patch there which basically does the same thing but in `wxDVC` code only and so doesn't help solving this problem everywhere else.

@a-wi @MaartenBent and anybody else who has ever worked with wxMSW `wxDC` and related code: could you please review the changes here and let me know if you see any (actual or potential) problems with doing this? TIA!

vadz committed:
Increase maximum allowed size of test virtual grid in the sample

This allows to see that the grid is currently not redrawn correctly
under MSW when it has more than ~6,000,000 rows due to overflowing the
device space span of 2^27.

vadz committed:
Increase usable scrolling range in wxMSW by a factor of 10,000

Using GDI functions for the device origin translation limits the
scrolling range to 2^27 size of the device space in the controls using
PrepareDC(), as it works by adjusting the device origin, and this may be
too small when having many (millions) of rows, which is perfectly
possible with wxDataViewCtrl or wxGrid, for example.

So handle DC device origin translations ourselves in wxMSW, i.e. offset
the coordinates by device origin before passing them to the native
functions as this allows to use the full 32-bit range.

Closes #17550.

vadz commented:
Sorry, I stupidly forgot to rerun the tests before pushing, I'll fix the clipping box calculation a.s.a.p. (but I have some troubles understanding what's going on in the RTL case here, so it will take a bit longer).

MaartenBent commented:
I've never really used this functionality myself, but tried moving the origin around in the drawing sample pages and that seems to works fine.

vadz commented:
I think I've finally got this right, but review would still be welcome as it's really simple to make mistakes in this code (and hence it's especially great to have such detailed unit tests for it, thanks @a-wi for writing them!).

FWIW I've also considered making this change conditional or optional, i.e. still call `::SetViewportOrgEx()` if we can (i.e. the offset is small enough) or if some option is on (or off), but finally decided against it, as it would make the code even more complicated as we'd need to know whether we're using our origin offset or MSW one, and either our code works and then I don't really see any benefit in using MSW support for this, or it doesn't -- and then it needs to be fixed.

vadz committed:
Don't use array of a single POINT in wxMSW wxDC code

No real changes, just use a simple POINT rather than POINT[1].

vadz committed:
Fix wxDC::DeviceToLogical() in presence of transform matrix

It's not enough to just shift the result by the origin shift if there is
a transform matrix, we need to apply the reverse shift transformed by
this matrix in general.

This is not very efficient as we redo the same matrix operations on
every call, but using both device origin shift and transform matrix
together should be quite rare, so it's not clear if it's worth
optimizing this.

vadz commented:
OK, I think I've finally *really* got this right, after fixing the origin shift handling in presence of non-trivial transform matrix. I don't think there is any simpler way to do it correctly than what I did in the last commit, but @a-wi please let me know if I'm missing something here.

vadz committed:
Add a workaround for failing test under Wine

Debugging confirms that DPtoLP() simply returns wrong (i.e. different
from that returned under actual MSW) result when using Wine, so just
account for it in the test as it seems to be better than just skipping
the test entirely under Wine and there doesn't seem to be anything else
to do.

vadz commented:
As there hasn't been any feedback, I'm going to merge this to get some testing for it before 3.1.7.
